common/achievement/evaluate.ts

```diff
@@ -10,17 +19,62 @@ import { achievement_event } from '@prisma/client';
 
 const logger = getLogger('Achievement');
 
-export const evaluateAchievement = tracer.wrap('achievement.evaluateAchievement', _evaluateAchievement);
+interface IsConditionMetResult {
+    isConditionMet: boolean;
+    // A sum of all data aggregation field values
+    // Usually used for streaks or tiered achievements to show the current progress
+    aggregationResult: number;
+    // Is only set for streak achievements
+    shouldInvalidateStreak: boolean | null;
+}
+
+export const isAchievementConditionMet = tracer.wrap('achievement.isAchievementConditionMet', _isAchievementConditionMet);
+async function _isAchievementConditionMet(achievement: AchievementToCheck): Promise<IsConditionMetResult> {
+    const {
+        userId,
+        recordValue,
+        context,
+        relation,
+        template: { condition, conditionDataAggregations, type },
+    } = achievement;
+    // If skipAchievementBucketsBefore is set, it indicates that a user may have joined an already running course.
+    // Therefore, we should skip the achievement buckets before the user's join, so that he still has the possibility to achieve the achievements.
+    const skipAchievementBucketsBefore = context && context['skipAchievementBucketsBefore'] ? new Date(context['skipAchievementBucketsBefore']) : undefined;
+    // For streaks, we need to ignore future buckets, as they will interrupt the streak even though the user had no opportunity to complete them.
+    // Nevertheless, we still want to include future buckets if they are associated with an event. This occurs if you join a meeting before its official start time.
+    const skipAchievementBucketsAfter = type === AchievementType.STREAK ? new Date() : undefined;
+    const result = await evaluateAchievement(
+        userId,
+        condition,
+        conditionDataAggregations as ConditionDataAggregations,
+        recordValue,
+        relation,
+        skipAchievementBucketsBefore,
+        skipAchievementBucketsAfter
+    );
+
+    const aggrResult = Object.keys(conditionDataAggregations).reduce((acc, key) => acc + result.resultObject[key], 0);
```

Comment:
Are we certain that every key of conditionDataAggregations is in resultObject? Maybe add an assertion that checks for NaN?

Reply:
Good Idea 👍